### PR TITLE
Governance Origins via XCM

### LINKED ISSUE #4: Design and Implement Governance System
Statemint should be fully aligned with the Relay Chain to which it is connected. This will require the ability to define `Origin`s as coming from another chain (i.e. an implementation of `EnsureOneOf` that allows declaring another chain's origin).

Edit: Decisions based on comments:
- Statemint Root = Relay Chain Root
- Statemint ForceOrigin = More than half of Relay Chain Council
- Statemint Transaction Payment Origin = More than half of Relay Chain Council

------------------------------------------------
We also need to make some concrete decisions about these origins (i.e. Relay Chain root? Some fraction of Council/Technical Committee?):

- `Root` origin (ability to perform runtime upgrades)
- `ForceOrigin` origin (ability to perform `force_*` actions in `pallet-assets`)
- Origin to set transaction payment DOT to asset ratio (#5 )

In addition, should this have the Democracy pallet to allow users to make Statemint-specific proposals for the Relay Chain to decide on? Or should those proposals be made through the Relay Chain's governance?

### LINKED ISSUE #66: Fix XCM Weights
TODO here: https://github.com/paritytech/statemint/blob/dbd1155/runtime/statemine/src/lib.rs#L494

> This should use `UsingComponents`. You should be able to figure it out from my work on the cumulus test parachain.

### PR DESCRIPTION
**Pull request by apopiak**
Adds support for asset pallet `ForceOrigin` via relay chain governance.

related to #3 

closes #4
closes #66

### CODE REVIEW

**Comment by apopiak** [comment]
based on #49

**Review by joepetrowski — COMMENTED** [review]
Need to add `UpdateOrigin` to be the same as `ForceOrigin`

File: runtime/statemint/src/lib.rs [hunk 1]
```diff
@@ -324,16 +325,24 @@ parameter_types! {
...
+	EnsureXcm<IsMajorityOfBody<RelayLocation, UnitBody>>,
+>;
+
 impl pallet_assets::Config for Runtime {
 	type Event = Event;
 	type Balance = Balance;
 	type AssetId = u32;
 	type Currency = Balances;
 	// TODO: Change to proportion at least 60% (3/5) of Relay Chain Council.
 	// https://github.com/paritytech/statemint/issues/4
```
Comment (joepetrowski) [inline runtime/statemint/src/lib.rs:344]: Remove TODO

File: runtime/statemint/src/lib.rs [hunk 2]
```diff
@@ -324,16 +325,22 @@ parameter_types! {
 	// https://github.com/paritytech/substrate/blob/069917b/frame/assets/src/lib.rs#L257L271
 	pub const MetadataDepositBase: Balance = deposit(1, 68);
 	pub const MetadataDepositPerByte: Balance = deposit(0, 1);
+	pub const UnitBody: BodyId = BodyId::Unit;
 }
 
+// We allow root and the relay council to execute privileged asset operations.
+pub type ForceOrigin =  EnsureOneOf<
```
Comment (kianenigma) [inline runtime/statemint/src/lib.rs:332]:
```suggestion
pub type AssetsForceOrigin =  EnsureOneOf<
```

File: runtime/statemint/src/lib.rs [hunk 3]
```diff
@@ -324,16 +325,22 @@ parameter_types! {
...
+
 impl pallet_assets::Config for Runtime {
 	type Event = Event;
 	type Balance = Balance;
 	type AssetId = u32;
 	type Currency = Balances;
-	// TODO: Change to proportion at least 60% (3/5) of Relay Chain Council.
-	// https://github.com/paritytech/statemint/issues/4
-	type ForceOrigin = frame_system::EnsureRoot<AccountId>;
+	type ForceOrigin = ForceOrigin;
```
Comment (kianenigma) [inline runtime/statemint/src/lib.rs:343]:
```suggestion
	type ForceOrigin = AssetsForceOrigin;
```

File: runtime/statemine/src/lib.rs [hunk 4]
```diff
@@ -651,18 +666,16 @@ parameter_types! {
...
 }
 
-frame_support::ord_parameter_types! {
-	pub const RelayChainCouncilOrigin: AccountId = <
-		frame_support::PalletId
-		as
-		AccountIdConversion<AccountId>
-	>::into_account(&frame_support::PalletId(*b"TODOTODO"));
-}
+pub type UpdateOrigin = EnsureOneOf<
```
Comment (kianenigma) [inline runtime/statemine/src/lib.rs:669]:
```suggestion
pub type CollatorSelectionUpdateOrigin = EnsureOneOf<
```

File: runtime/statemine/src/lib.rs [hunk 5]
```diff
@@ -651,18 +666,16 @@ parameter_types! {
...
+	AccountId,
+	EnsureRoot<AccountId>,
+	EnsureXcm<IsMajorityOfBody<RelayLocation, UnitBody>>,
+>;
 
 impl pallet_collator_selection::Config for Runtime {
 	type Event = Event;
 	type Currency = Balances;
-	type UpdateOrigin = frame_system::EnsureSignedBy<RelayChainCouncilOrigin, AccountId>;
+	type UpdateOrigin = UpdateOrigin;
```
Comment (kianenigma) [inline runtime/statemine/src/lib.rs:678]:
```suggestion
	type UpdateOrigin = CollatorSelectionUpdateOrigin;
```

File: runtime/statemint/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply runtime/statemint/src/lib.rs:332]: Might have multiple pallets that accept a `type ForceOrigin` etc.

File: runtime/statemine/src/lib.rs [hunk 6]
```diff
@@ -323,16 +324,22 @@ parameter_types! {
 	// https://github.com/paritytech/substrate/blob/069917b/frame/assets/src/lib.rs#L257L271
 	pub const MetadataDepositBase: Balance = deposit(1, 68);
 	pub const MetadataDepositPerByte: Balance = deposit(0, 1);
+	pub const UnitBody: BodyId = BodyId::Unit;
 }
 
+// We allow root and the relay council to execute privileged asset operations.
```
Comment (joepetrowski) [inline runtime/statemine/src/lib.rs:330]:
```suggestion
/// We allow root and the Relay Chain council to execute privileged asset operations.
```

File: runtime/statemint/src/lib.rs [hunk 7]
```diff
@@ -324,16 +325,22 @@ parameter_types! {
 	// https://github.com/paritytech/substrate/blob/069917b/frame/assets/src/lib.rs#L257L271
 	pub const MetadataDepositBase: Balance = deposit(1, 68);
 	pub const MetadataDepositPerByte: Balance = deposit(0, 1);
+	pub const UnitBody: BodyId = BodyId::Unit;
 }
 
+// We allow root and the relay council to execute privileged asset operations.
```
Comment (joepetrowski) [inline runtime/statemint/src/lib.rs:331]:
```suggestion
/// We allow root and the Relay Chain council to execute privileged asset operations.
```

File: runtime/statemint/src/lib.rs [hunk 8]
```diff
@@ -656,18 +671,16 @@ parameter_types! {
...
 }
 
-frame_support::ord_parameter_types! {
-	pub const RelayChainCouncilOrigin: AccountId = <
-		frame_support::PalletId
-		as
-		AccountIdConversion<AccountId>
-	>::into_account(&frame_support::PalletId(*b"TODOTODO"));
-}
+pub type CollatorSelectionUpdateOrigin = EnsureOneOf<
```
Comment (joepetrowski) [inline runtime/statemint/src/lib.rs:674]:
```suggestion
/// We allow root and the Relay Chain council to execute privileged collator selection operations.
pub type CollatorSelectionUpdateOrigin = EnsureOneOf<
```

File: runtime/statemine/src/lib.rs [hunk 9]
```diff
@@ -651,18 +666,16 @@ parameter_types! {
...
 }
 
-frame_support::ord_parameter_types! {
-	pub const RelayChainCouncilOrigin: AccountId = <
-		frame_support::PalletId
-		as
-		AccountIdConversion<AccountId>
-	>::into_account(&frame_support::PalletId(*b"TODOTODO"));
-}
+pub type CollatorSelectionUpdateOrigin = EnsureOneOf<
```
Comment (joepetrowski) [inline runtime/statemine/src/lib.rs:669]:
```suggestion
/// We allow root and the Relay Chain council to execute privileged collator selection operations.
pub type CollatorSelectionUpdateOrigin = EnsureOneOf<
```

File: runtime/statemine/src/lib.rs [hunk 10]
```diff
@@ -323,16 +324,22 @@ parameter_types! {
 	// https://github.com/paritytech/substrate/blob/069917b/frame/assets/src/lib.rs#L257L271
 	pub const MetadataDepositBase: Balance = deposit(1, 68);
 	pub const MetadataDepositPerByte: Balance = deposit(0, 1);
+	pub const UnitBody: BodyId = BodyId::Unit;
```
Comment (joepetrowski) [inline runtime/statemine/src/lib.rs:327]: I think this was set to `Unit` in the context of Rococo, but on Polkadot/Kusama, we want it to be the Council.
```suggestion
	pub const UnitBody: BodyId = BodyId::Executive;
```

File: runtime/statemint/src/lib.rs [hunk 11]
```diff
@@ -324,16 +325,22 @@ parameter_types! {
 	// https://github.com/paritytech/substrate/blob/069917b/frame/assets/src/lib.rs#L257L271
 	pub const MetadataDepositBase: Balance = deposit(1, 68);
 	pub const MetadataDepositPerByte: Balance = deposit(0, 1);
+	pub const UnitBody: BodyId = BodyId::Unit;
```
Comment (joepetrowski) [inline runtime/statemint/src/lib.rs:328]:
```suggestion
	pub const UnitBody: BodyId = BodyId::Executive;
```

File: runtime/statemine/src/lib.rs (hunk 10 above)
Comment (riusricardo) [reply runtime/statemine/src/lib.rs:327]: I think that `Unit` this is still valid for Council. In any case I want to suggest to make the change in a follow up PR as the Aura PR will bring many changes and I want to merge this.

**Review by riusricardo — APPROVED** [review]
LGTM.

File: runtime/statemine/src/lib.rs (hunk 10 above)
Comment (gavofyork) [reply runtime/statemine/src/lib.rs:261]: Should be Executive. Unit is not valid.

**Pull request merged**